app/src/main/kotlin/com/gdgnantes/devfest/android/SyncService.kt

```diff
@@ -0,0 +1,147 @@
+import com.gdgnantes.devfest.android.model.Schedule
+import com.gdgnantes.devfest.android.model.toContentValues
+import com.gdgnantes.devfest.android.provider.ScheduleContract
+import com.google.firebase.crash.FirebaseCrash
+import retrofit2.Call
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import retrofit2.http.GET
+import retrofit2.http.Header
+import java.io.IOException
+import java.lang.RuntimeException
+import java.util.concurrent.TimeUnit
+
+class SyncService : IntentService(TAG) {
+
+    // NOTE Cyril
+    // The current implementation of the sync mechanism is fairly simple …
+    // not to say stupid^^.
+    // The sync consists on heavily relying on the server response code:
+    //   - a 304 response indicates the local schedule is already up-to-date
+    //     and, as a consequence, nothing needs to be done locally. In that
+    //     case, the sync stops immediately;
+    //   - a 200 response code indicates the local schedule is outdated. The
+    //     code consists on removing everything that was persisted locally
+    //     and insert the newly retrieved content from the server.
+    //   - other response codes are treated as errors
+    //
+    // Obviously, this algorithm could be improved greatly by doing diffs
+    // which would avoid unnecessary insert/delete operations. The main
+    // /advantage of the current implementation is it keeps the code as
+    // simple as possible.
+
+    companion object {
+        private const val TAG = "SyncService"
+
+        private val AMOUNT = TimeUnit.MINUTES.toMillis(10)
+
+        private var lastSync = -1L
+
+        fun sync(context: Context) {
+            if (lastSync > 0 && (System.currentTimeMillis() - lastSync) <= AMOUNT) {
+                // No need to sync now as it's been done recently
+                // Let's skip the request
+                return
+            }
+            context.startService(Intent(context, SyncService::class.java))
+        }
+    }
+
+    interface ScheduleApi {
+
+        interface Headers {
+            companion object {
+                const val ETAG = "etag"
+                const val IF_NONE_MATCH = "If-None-Match"
+            }
+        }
+
+        interface ResponseCodes {
+            companion object {
+                const val OK = 200
+                const val NO_CONTENT = 304
+            }
+        }
+
+        @GET("schedule")
+        fun getSchedule(@Header(Headers.IF_NONE_MATCH) etag: String): Call<Schedule>
+    }
+
+    override fun onHandleIntent(intent: Intent) {
+        val api: ScheduleApi = Retrofit.Builder()
+                .baseUrl(AppConfig.ENDPOINT)
+                .addConverterFactory(GsonConverterFactory.create(JsonConverters.main))
+                .build()
+                .create(ScheduleApi::class.java)
+
+        try {
+            val prefs = PreferencesManager.from(this)
+            val etag = prefs.scheduleETag ?: AppConfig.SEED_ETAG
```

Comment:
To keep things simple, we'll consider we can't update the schedule with an app update. Seed is only used once at the beginning and that's all. After that, updates are done via the API.